### PR TITLE
Refactor DepositCharge so it can be called from ImpactX.

### PR DESCRIPTION
**Pull request by atmyers**
Todo: 
- [x] Handle static members / methods of WarpX.

### CODE REVIEW

**Comment by PhilMiller** [comment]
You may want to base this on #2550, which is basically ready for review and integration.

File: Source/Particles/WarpXParticleContainer.cpp [hunk 1]
```diff
@@ -598,146 +598,56 @@ WarpXParticleContainer::DepositCharge (WarpXParIter& pti, RealVector& wp,
...
+        const int nc = WarpX::ncomps;
+
+        // Get tile box where charge is deposited.
+        // The tile box is different when depositing in the buffers (depos_lev<lev)
+        // or when depositing inside the level (depos_lev=lev)
+        amrex::Box tilebox;
+        if (lev == depos_lev) {
+            tilebox = pti.tilebox();
+        } else {
+            tilebox = amrex::coarsen(pti.tilebox(), ref_ratio);
```
Comment (ax3l) [inline Source/Particles/WarpXParticleContainer.cpp:621]:
```suggestion
            ref_ratio = WarpX::RefRatio(depos_lev);
            tilebox = amrex::coarsen(pti.tilebox(), ref_ratio);
```

File: Source/Particles/WarpXParticleContainer.cpp [hunk 2]
```diff
@@ -598,146 +598,56 @@ WarpXParticleContainer::DepositCharge (WarpXParIter& pti, RealVector& wp,
...
-            m_v_galilean[0]*time_shift_rho_new,
-            m_v_galilean[1]*time_shift_rho_new,
-            m_v_galilean[2]*time_shift_rho_new };
-    }
-    const std::array<Real, 3>& xyzmin = WarpX::LowerCorner(tilebox, galilean_shift, depos_lev);
+    if (!do_not_deposit) {
+        WarpX& warpx = WarpX::GetInstance();
+        const amrex::IntVect& ng_rho = warpx.get_ng_depos_rho();
+        const std::array<amrex::Real,3>& dx = WarpX::CellSize(std::max(depos_lev,0));
+        const amrex::IntVect& ref_ratio = WarpX::RefRatio(depos_lev);
```
Comment (ax3l) [inline Source/Particles/WarpXParticleContainer.cpp]:
```suggestion
        amrex::IntVect ref_ratio = amrex::IntVect(AMREX_D_DECL(1, 1, 1));
```

File: Source/Particles/Deposition/ChargeDeposition.H [hunk 3]
```diff
@@ -213,4 +214,124 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
 #endif
 }
 
+template <typename PC>
```
Comment (ax3l) [inline Source/Particles/Deposition/ChargeDeposition.H]: For later:
- namespace
- own file
- doxygen string

File: Source/Particles/Deposition/ChargeDeposition.H [hunk 4]
```diff
@@ -213,4 +214,124 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
...
+void DepositChargeDoIt (typename PC::ParIterType& pti,
```
Comment (ax3l) [inline Source/Particles/Deposition/ChargeDeposition.H]: Somewhere in this API, `WarpX::do_device_synchronize` is called (probably in a lower-level call).

We need to factor this out as an option, too.

Comment (atmyers) [reply Source/Particles/Deposition/ChargeDeposition.H]: It's coming from the WarpXProfilerWrapper.

File: CMakeLists.txt [hunk 5]
```diff
@@ -191,6 +191,7 @@ if(WarpX_APP)
     target_sources(app PRIVATE Source/main.cpp)
 endif()
 
+#add_subdirectory(Source/ABLASTR)
```
Comment (ax3l) [inline CMakeLists.txt:198]: once we rebase/merge as #2263 is merged, we can add the extra include in the `# own headers` part above

File: Source/ABLASTR/ABLASTR.H [hunk 6]
```diff
@@ -0,0 +1,179 @@
+/* Copyright 2019 Axel Huebl, Andrew Myers
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef ABLASTR_H_
```
Comment (ax3l) [inline Source/ABLASTR/ABLASTR.H]: Some thoughts:
- maybe let us name the file `DepositCharge.H`
- use lowercase in `namespace ablastr`

File: CMakeLists.txt (hunk 5 above)
Comment (ax3l) [reply CMakeLists.txt:198]: currently same root with intentional directory prefix. keep as is for now.

**Review by ax3l — APPROVED** [review]
:tada:

File: Source/ABLASTR/DepositCharge.H [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    // On GPU: particles deposit directly on the rho array, which usually have a larger number of guard cells
+#ifndef AMREX_USE_GPU
+    const amrex::IntVect range = ng_rho - shape_extent;
+#else
+    const amrex::IntVect range = rho->nGrowVect() - shape_extent;
+#endif
+
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+        amrex::numParticlesOutOfRange(pti, range) == 0,
+        "Particles shape does not fit within tile (CPU) or guard cells (GPU) used for charge deposition");
```
Comment (PhilMiller) [inline Source/ABLASTR/DepositCharge.H:107]: Is it really worth unconditionally doing an extra pass over all of the particles for this error check? Could the same check be made an assertion inside the kernel that loops over the particles anyway?

Comment (atmyers) [reply Source/ABLASTR/DepositCharge.H:107]: I tend to agree with you that this is too expensive to always do - FTR this was already the behavior of DepositCharge, not new to this PR. Maybe this is was useful when developing the large time step algorithm, though, @RemiLehe? 

Perhaps we should do this only in DEBUG builds...

Comment (ax3l) [reply Source/ABLASTR/DepositCharge.H:107]: It's indeed not free but also the first line of defense where we see issues when implementing new algorithms or (over-)optimizing our comms. (Contrary to chasing NaNs & segfaults down the road.)

Note that we cannot afford full debug builds in CI atm., so we might not catch a new issues immediately if that one is debug only. But (debug || a runtime option) or so would work. Or reacting/adding this on (DEBUG || AMReX_TESTING/TEST=TRUE || AMReX_ASSERTIONS/USE_ASSERTION=TRUE) - the latter two are also ON in CI.
In that case, `AMREX_ALWAYS_ASSERT_WITH_MESSAGE` -> `AMREX_ASSERT_WITH_MESSAGE`?

This applies to both charge and current deposition.

Comment (atmyers) [reply Source/ABLASTR/DepositCharge.H:107]: `AMREX_ASSERT_WITH_MESSAGE` will assert if `USE_ASSERTION=TRUE`, even if `DEBUG=FALSE`, so yes I think we should make this `AMREX_ASSERT_WITH_MESSAGE`.

Comment (RemiLehe) [reply Source/ABLASTR/DepositCharge.H:107]: OK, yes, I agree that this assertion should only be done in `DEBUG` mode, for performance reasons.

Comment (atmyers) [reply Source/ABLASTR/DepositCharge.H:107]: PR to do this here: https://github.com/ECP-WarpX/WarpX/pull/2688

**Pull request merged**